src/background/tab-state.js

```diff
@@ -140,15 +143,47 @@ export default function TabState(initialState, onchange) {
   };
 
   /**
-   * Request the current annotation count for the tab's URL
+   * Request the current annotation count for the tab's URL.
+   * It debounces previous requests within the value defined by
+   * WAITMILLISECONDS
    *
    * @method
    * @param {integer} tabId The id of the tab.
    * @param {string} tabUrl The URL of the tab.
    */
   this.updateAnnotationCount = function (tabId, tabUrl) {
-    return uriInfo.getAnnotationCount(tabUrl).then(count => {
-      this.setState(tabId, { annotationCount: count });
+    const WAITMILLISECONDS = 3000; //
+    const pendingRequest = pendingAnnotationCountRequests.get(tabId);
+
+    // Cancel pending requests for the specific tabId, if any
+    if (pendingRequest && pendingRequest.cancel) {
+      pendingRequest.cancel();
+    }
+
+    const fetchCount = async () => {
+      try {
+        const annotationCount = await uriInfo.getAnnotationCount(tabUrl);
+        this.setState(tabId, { annotationCount });
+      } finally {
+        // Clear the pending request
+        pendingAnnotationCountRequests.delete(tabId);
```

Comment:
What happens if this method (`updateAnnotationCount`) is invoked again before this Promise settles (i.e. before the `pendingAnnotationCountRequests` map is updated? 

That is, what if the `timeout` has fired, this request is in-flight (and thus canceling the timeout won't do anything), and `updateAnnotationCount` is called?

Reply:
In modern browsers [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) can be used to cancel `fetch` requests. If we wanted to use that API we could either bump the minimum Chrome version to one that supports it or just not implement cancellation in older versions of Chrome.

> That's why the cancelling function also rejects the pending promise.

I don't see any promise rejection happening in this code. The `cancel` callback clears the timeout and resolves it with a count of 0.